azure-spring-boot/src/main/java/com/microsoft/azure/keyvault/spring/KeyVaultOperation.java

```diff
@@ -33,68 +37,84 @@ public KeyVaultOperation(final KeyVaultClient keyVaultClient, final String vault
     public KeyVaultOperation(final KeyVaultClient keyVaultClient, String vaultUri, final long refreshInterval) {
         this.cacheRefreshIntervalInMs = refreshInterval;
         this.keyVaultClient = keyVaultClient;
+        // TODO(pan): need to validate why last '/' need to be truncated.
+        this.vaultUri = StringUtils.trimTrailingCharacter(vaultUri.trim(), '/');
 
-        vaultUri = vaultUri.trim();
-        if (vaultUri.endsWith("/")) {
-            vaultUri = vaultUri.substring(0, vaultUri.length() - 1);
-        }
-        this.vaultUri = vaultUri;
-
-        createOrUpdateHashMap();
+        fillSecretsHashMap();
     }
 
     public String[] list() {
         try {
             this.rwLock.readLock().lock();
-            return Collections.list(this.propertyNamesHashMap.keys())
-                    .toArray(new String[this.propertyNamesHashMap.size()]);
+            return Collections.list(propertyNamesHashMap.keys()).toArray(new String[propertyNamesHashMap.size()]);
         } finally {
             this.rwLock.readLock().unlock();
         }
     }
 
-    public Object get(final String secretName) {
-        // NOTE: azure keyvault secret name convention: ^[0-9a-zA-Z-]+$ "." is not allowed
-        final String localSecretName = secretName.replace(".", "-");
+    private String getKeyvaultSecretName(@NonNull String property) {
+        if (property.matches("[a-z0-9A-Z-]+")) {
+            return property.toLowerCase(Locale.US);
+        } else if (property.matches("[A-Z0-9_]+")) {
+            return property.toLowerCase(Locale.US).replaceAll("_", "-");
+        } else {
+            return property.toLowerCase(Locale.US)
+                    .replaceAll("-", "")     // my-project -> myproject
+                    .replaceAll("_", "")     // my_project -> myproject
+                    .replaceAll("\\.", "-"); // acme.myproject -> acme-myproject
+        }
+    }
+
+    /**
+     * For convention we need to support all relaxed binding format from spring, these may include:
+     * <table>
+     * <tr><td>Spring relaxed binding names</td></tr>
+     * <tr><td>acme.my-project.person.first-name</td></tr>
+     * <tr><td>acme.myProject.person.firstName</td></tr>
+     * <tr><td>acme.my_project.person.first_name</td></tr>
+     * <tr><td>ACME_MYPROJECT_PERSON_FIRSTNAME</td></tr>
+     * </table>
+     * But azure keyvault only allows ^[0-9a-zA-Z-]+$ and case insensitive, so there must be some conversion
+     * between spring names and azure keyvault names.
+     * For example, the stated below 4 properties should be convert to acme-myproject-person-firstname in keyvault.
```

Comment:
Comment line 79 move to up? There is no `below 4 properties`.

Reply:
Oh, typo, I mean above here, will fix it ASAP.